src/common/buffer.cc

```diff
@@ -1501,39 +1495,109 @@ using namespace ceph;
   void buffer::list::append(char c)
   {
     // put what we can into the existing append_buffer.
-    unsigned gap = append_buffer.unused_tail_length();
+    unsigned gap = get_append_buffer_unused_tail_length();
     if (!gap) {
-      // make a new append_buffer!
-      append_buffer = raw_combined::create(CEPH_BUFFER_APPEND_SIZE, 0,
-					   get_mempool());
-      append_buffer.set_length(0);   // unused, so far.
-    }
-    append(append_buffer, append_buffer.append(c) - 1, 1);	// add segment to the list
+      // make a new buffer!
+      auto buf = ptr_node::create(
+	raw_combined::create(CEPH_BUFFER_APPEND_SIZE, 0, get_mempool()));
+      buf->set_length(0);   // unused, so far.
+      _carriage = buf.get();
+      _buffers.push_back(*buf.release());
+    } else if (unlikely(_carriage != &_buffers.back())) {
+      auto bptr = ptr_node::create(*_carriage, _carriage->length(), 0);
+      _carriage = bptr.get();
+      _buffers.push_back(*bptr.release());
+    }
+    _carriage->append(c);
+    _len++;
+  }
+
+  buffer::ptr buffer::list::always_empty_bptr;
```

Comment:
if the only purpose of `always_empty_bptr` is to be assigned to `_carriage` as a sentry. why not just use `nullptr` instead?

Reply:
The idea was to avoid extra costs (conditionals, jumps etc.) in very hot calls like `get_append_buffer_unused_tail_length()`.